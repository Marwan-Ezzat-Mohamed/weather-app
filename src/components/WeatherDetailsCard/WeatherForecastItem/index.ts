import WeatherForecastItem from "./WeatherForecastItem";

export { WeatherForecastItem };
