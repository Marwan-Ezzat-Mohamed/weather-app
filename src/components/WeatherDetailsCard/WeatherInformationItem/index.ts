import WeatherInformationItem from "./WeatherInformationItem";

export { WeatherInformationItem };
