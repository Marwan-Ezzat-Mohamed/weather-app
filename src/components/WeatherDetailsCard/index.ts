import WeatherDetailsCard from "./WeatherDetailsCard";

export { WeatherDetailsCard };
