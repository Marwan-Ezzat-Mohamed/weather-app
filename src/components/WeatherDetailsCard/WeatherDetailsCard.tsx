import { useMemo } from "react";
import { FaWind } from "react-icons/fa";
import { MdWaterDrop } from "react-icons/md";
import { WiBarometer } from "react-icons/wi";
import { AiFillStar, AiOutlineStar } from "react-icons/ai";
import { TiArrowUpThick, TiArrowDownThick } from "react-icons/ti";
import { WeatherForecastItem } from "./WeatherForecastItem";
import { WeatherInformationItem } from "./WeatherInformationItem";
import WeatherIcon from "@components/WeatherIcon";
import { City } from "@/types";
import { HOUR, formatDate } from "@/utils";
import { WeatherHour } from "@/services/types";

interface IWeatherDetailsCardProps {
  city: City;
  temp: number;
  humidity: number;
  wind: number;
  feelsLike: number;
  maxTemp: number;
  minTemp: number;
  pressure: number;
  localTime: string;
  description: string;
  isFavorite: boolean;
  weatherConditionText: string;
  isDay: boolean;
  weatherHour: WeatherHour[];
  onFavoriteClick: () => void;
}

const FORECAST_HOURS = 6;
const WeatherDetailsCard = ({
  city,
  temp,
  localTime,
  weatherConditionText,
  isDay,
  humidity,
  wind,
  feelsLike,
  maxTemp,
  minTemp,
  pressure,
  description,
  isFavorite,
  weatherHour,
  onFavoriteClick,
}: IWeatherDetailsCardProps) => {
  const formattedDate = formatDate(localTime);

  const filteredWeatherHour = weatherHour.filter((w) => {
    //get the next 7 hours from now (including the current hour) and not the past ones
    const time = new Date(w.time).getTime();
    const now = new Date().getTime();
    return time >= now && time <= now + FORECAST_HOURS * HOUR;
  });

  const weatherInfoItems = useMemo(
    () => [
      {
        label: "Max",
        valueLabel: `${maxTemp.toFixed(0)}째`,
        Icon: TiArrowUpThick,
      },
      {
        label: "Min",
        valueLabel: `${minTemp.toFixed(0)}째`,
        Icon: TiArrowDownThick,
      },
      {
        label: "Wind",
        valueLabel: `${wind.toFixed(0)}kph`,
        Icon: FaWind,
      },
      {
        label: "Humidity",
        valueLabel: `${humidity.toFixed(0)}%`,
        Icon: MdWaterDrop,
      },
      {
        label: "Pressure",
        valueLabel: `${pressure.toFixed(0)}hPa`,
        Icon: WiBarometer,
      },
    ],
    [humidity, maxTemp, minTemp, pressure, wind]
  );

  return (
    <div className="bg-secondary relative flex  w-full justify-center rounded-2xl p-5 transition duration-200 hover:shadow-xl">
      <div
        className="absolute bottom-0 left-0 m-2 text-4xl text-yellow-500"
        onClick={onFavoriteClick}
        data-testid="favorite-icon"
      >
        {isFavorite ? <AiFillStar /> : <AiOutlineStar />}
      </div>
      <div className="w-full space-y-5">
        <div className="flex justify-between text-2xl">
          <span className="font-bold">{city.name}</span>
          <span className="font-semibold ">{formattedDate}</span>
        </div>
        <section className="flex flex-col items-center justify-around md:flex-row">
          <div className=" flex flex-col items-center  ">
            <div className="flex items-center  ">
              <WeatherIcon
                weatherConditionText={weatherConditionText}
                big={true}
                isDay={isDay}
              />

              <div className="flex items-center justify-center">
                <h1 className="text-8xl"> {temp.toFixed(0)}째C</h1>
              </div>
            </div>
            <span className="text-4xl font-semibold capitalize">
              {description}
            </span>
          </div>
          <div className="flex flex-col text-2xl">
            <div className="mb-5 flex items-center justify-between">
              <h1> Feels like {feelsLike.toFixed(0)}째 </h1>
            </div>
            {weatherInfoItems.map((item, index) => (
              <WeatherInformationItem
                key={index}
                label={item.label}
                valueLabel={item.valueLabel}
                Icon={item.Icon}
              />
            ))}
          </div>
        </section>
        <section className="flex flex-col gap-2">
          <h1 className="text-2xl font-semibold">
            Next {filteredWeatherHour.length} hours
          </h1>
          <div className="flex flex-wrap justify-center gap-5">
            {filteredWeatherHour?.map((w) => {
              const time = new Date(w.time).getTime();
              const formattedTime = new Intl.DateTimeFormat("en-US", {
                hour: "numeric",
                hour12: true,
              }).format(time);
              return (
                <WeatherForecastItem
                  weather={{
                    time: w.time,
                    conditionText: w.condition.text,
                    temp_c: w.temp_c,
                    isDay: !!w.is_day,
                  }}
                  formattedTime={formattedTime}
                  key={w.time}
                />
              );
            })}
          </div>
        </section>
      </div>
    </div>
  );
};

export default WeatherDetailsCard;
