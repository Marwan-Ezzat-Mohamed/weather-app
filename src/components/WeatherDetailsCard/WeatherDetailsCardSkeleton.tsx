import { WeatherInformationItemSkeleton } from "./WeatherInformationItem/WeatherInformationItemSkeleton";
import { WeatherForecastItemSkeleton } from "./WeatherForecastItem/WeatherForecastItemSkeleton";

const WeatherDetailsCardSkeleton = () => {
  return (
    <div className="flex w-full justify-center rounded-2xl bg-slate-100 p-5 transition duration-200 hover:shadow-2xl dark:bg-slate-800">
      <div className="w-full space-y-5">
        <div className="flex justify-between text-2xl">
          <div className="h-6 w-1/3 animate-pulse rounded-md bg-gray-300 dark:bg-gray-500 "></div>
          <div className="h-6 w-1/3 animate-pulse rounded-md bg-gray-300 dark:bg-gray-500"></div>
        </div>
        <section className="flex justify-around">
          <div className="flex flex-col items-center">
            <div className="flex items-center">
              <div className="h-24 w-24 animate-pulse rounded-full bg-gray-300 dark:bg-gray-500"></div>
              <div className="ml-4 h-32 w-32 animate-pulse rounded-md bg-gray-300 dark:bg-gray-500"></div>
            </div>
          </div>
          <div className="flex flex-col text-2xl">
            <div className="mb-5 flex justify-between gap-5">
              <div className="h-6 w-2/3 animate-pulse rounded-md bg-gray-300 dark:bg-gray-500"></div>
              <div className="h-6 w-1/3 animate-pulse rounded-md bg-gray-300 dark:bg-gray-500"></div>
            </div>
            <div className="flex flex-col gap-2">
              <WeatherInformationItemSkeleton />
              <WeatherInformationItemSkeleton />
              <WeatherInformationItemSkeleton />
              <WeatherInformationItemSkeleton />
            </div>
          </div>
        </section>
        <section className="flex flex-col gap-2">
          <span className="h-7 w-2/3 animate-pulse rounded bg-gray-400 dark:bg-gray-500"></span>

          <div className="flex flex-wrap justify-center gap-5">
            <WeatherForecastItemSkeleton />
            <WeatherForecastItemSkeleton />
            <WeatherForecastItemSkeleton />
            <WeatherForecastItemSkeleton />
            <WeatherForecastItemSkeleton />
            <WeatherForecastItemSkeleton />
          </div>
        </section>
      </div>
    </div>
  );
};

export default WeatherDetailsCardSkeleton;
