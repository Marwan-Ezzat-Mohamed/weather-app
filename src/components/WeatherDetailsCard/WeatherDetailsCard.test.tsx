import { render, screen, fireEvent } from "@testing-library/react";
import WeatherDetailsCard from "./WeatherDetailsCard";

const mockWeatherHour = [
  {
    time_epoch: 1630549200,
    time: "2021-09-02 00:00",
    temp_c: 20.0,
    temp_f: 68.0,
    is_day: 0,

    condition: {
      text: "Clear",
      icon: "//cdn.weatherapi.com/weather/64x64/night/113.png",
      code: 1000,
    },
    wind_mph: 4.5,
    wind_kph: 7.2,
    wind_degree: 0,
    wind_dir: "N",
    pressure_mb: 1014.0,
    pressure_in: 30.4,
    precip_mm: 0.0,
    precip_in: 0.0,
    humidity: 50,
    cloud: 0,
    feelslike_c: 22.0,
    feelslike_f: 71.6,
    windchill_c: 20.0,

    windchill_f: 68.0,
    heatindex_c: 22.0,
    heatindex_f: 71.6,
    dewpoint_c: 9.0,
    dewpoint_f: 48.2,
    will_it_rain: 0,
    chance_of_rain: 0,
    will_it_snow: 0,
    chance_of_snow: 0,
    vis_km: 10.0,
    vis_miles: 6.0,
    gust_mph: 6.9,
    gust_kph: 11.2,
    uv: 1.0,
  },
  // Add more sample data as needed
];

const sampleCity = {
  id: "1",
  name: "Sample City",
  lat: 123.45,
  lng: 67.89,
};

const mockProps = {
  city: sampleCity,
  temp: 20,
  localTime: "2023-09-20T15:30:00Z",
  weatherConditionText: "Clear",
  isDay: true,
  humidity: 50,
  wind: 10,
  feelsLike: 22,
  maxTemp: 30,
  minTemp: 15,
  pressure: 1010,
  description: "Clear sky",
  isFavorite: false,
  weatherHour: mockWeatherHour,
  onFavoriteClick: jest.fn(),
};

describe("WeatherDetailsCard Component", () => {
  it("matches snapshot", () => {
    const { container } = render(<WeatherDetailsCard {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it("displays city name", () => {
    render(<WeatherDetailsCard {...mockProps} />);
    const cityNameElement = screen.getByText(sampleCity.name);
    expect(cityNameElement).toBeInTheDocument();
  });

  it("handles favorite click", () => {
    const mockOnFavoriteClick = jest.fn();
    render(
      <WeatherDetailsCard
        {...mockProps}
        onFavoriteClick={mockOnFavoriteClick}
      />
    );
    const favoriteIcon = screen.getByTestId("favorite-icon");

    fireEvent.click(favoriteIcon);
    expect(mockOnFavoriteClick).toHaveBeenCalled();
  });

  // Add more test cases to cover other component features
});
