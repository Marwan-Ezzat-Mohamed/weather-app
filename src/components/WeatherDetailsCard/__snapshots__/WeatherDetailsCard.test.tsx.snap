// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherDetailsCard Component matches snapshot 1`] = `
<div>
  <div
    class="relative flex w-full  justify-center rounded-2xl bg-slate-100 p-5 transition duration-200 hover:shadow-2xl dark:bg-slate-800"
  >
    <div
      class="absolute bottom-0 left-0 m-2 text-4xl text-yellow-500"
      data-testid="favorite-icon"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3zM664.8 561.6l36.1 210.3L512 672.7 323.1 772l36.1-210.3-152.8-149L417.6 382 512 190.7 606.4 382l211.2 30.7-152.8 148.9z"
        />
      </svg>
    </div>
    <div
      class="w-full space-y-5"
    >
      <div
        class="flex justify-between text-2xl"
      >
        <span
          class="font-bold"
        >
          Sample City
        </span>
        <span
          class="font-semibold "
        >
          Wed 5:30 PM
        </span>
      </div>
      <section
        class="flex justify-around flex-col items-center md:flex-row"
      >
        <div
          class=" flex flex-col items-center  "
        >
          <div
            class="flex items-center  "
          >
            <canvas
              height="140"
              width="140"
            />
            <div
              class="flex items-center justify-center"
            >
              <h1
                class="text-8xl"
              >
                 
                20
                째C
              </h1>
            </div>
          </div>
          <span
            class="text-4xl font-semibold capitalize"
          >
            Clear sky
          </span>
        </div>
        <div
          class="flex flex-col text-2xl"
        >
          <div
            class="mb-5 flex items-center justify-between"
          >
            <h1>
               Feels like 
              22
              째 
            </h1>
          </div>
          <div
            class="flex justify-start"
          >
            <div
              class="flex w-40 items-center gap-2 text-center text-3xl"
            >
              <svg
                baseProfile="tiny"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                version="1.2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 3.172l-6.414 6.414c-.781.781-.781 2.047 0 2.828s2.047.781 2.828 0l1.586-1.586v7.242c0 1.104.895 2 2 2 1.104 0 2-.896 2-2v-7.242l1.586 1.586c.391.391.902.586 1.414.586s1.023-.195 1.414-.586c.781-.781.781-2.047 0-2.828l-6.414-6.414z"
                />
              </svg>
              <span
                class="text-2xl"
              >
                Max
              </span>
            </div>
            <h1
              class="text-2xl"
            >
               
              30째
               
            </h1>
          </div>
          <div
            class="flex justify-start"
          >
            <div
              class="flex w-40 items-center gap-2 text-center text-3xl"
            >
              <svg
                baseProfile="tiny"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                version="1.2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M18.414 10.656c-.781-.781-2.047-.781-2.828 0l-1.586 1.586v-7.242c0-1.105-.896-2-2-2-1.105 0-2 .895-2 2v7.242l-1.586-1.586c-.781-.781-2.047-.781-2.828 0s-.781 2.047 0 2.828l6.414 6.414 6.414-6.414c.781-.781.781-2.046 0-2.828z"
                />
              </svg>
              <span
                class="text-2xl"
              >
                Min
              </span>
            </div>
            <h1
              class="text-2xl"
            >
               
              15째
               
            </h1>
          </div>
          <div
            class="flex justify-start"
          >
            <div
              class="flex w-40 items-center gap-2 text-center text-3xl"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M156.7 256H16c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h142.2c15.9 0 30.8 10.9 33.4 26.6 3.3 20-12.1 37.4-31.6 37.4-14.1 0-26.1-9.2-30.4-21.9-2.1-6.3-8.6-10.1-15.2-10.1H81.6c-9.8 0-17.7 8.8-15.9 18.4 8.6 44.1 47.6 77.6 94.2 77.6 57.1 0 102.7-50.1 95.2-108.6C249 291 205.4 256 156.7 256zM16 224h336c59.7 0 106.8-54.8 93.8-116.7-7.6-36.2-36.9-65.5-73.1-73.1-55.4-11.6-105.1 24.9-114.9 75.5-1.9 9.6 6.1 18.3 15.8 18.3h32.8c6.7 0 13.1-3.8 15.2-10.1C325.9 105.2 337.9 96 352 96c19.4 0 34.9 17.4 31.6 37.4-2.6 15.7-17.4 26.6-33.4 26.6H16c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16zm384 32H243.7c19.3 16.6 33.2 38.8 39.8 64H400c26.5 0 48 21.5 48 48s-21.5 48-48 48c-17.9 0-33.3-9.9-41.6-24.4-2.9-5-8.7-7.6-14.5-7.6h-33.8c-10.9 0-19 10.8-15.3 21.1 17.8 50.6 70.5 84.8 129.4 72.3 41.2-8.7 75.1-41.6 84.7-82.7C526 321.5 470.5 256 400 256z"
                />
              </svg>
              <span
                class="text-2xl"
              >
                Wind
              </span>
            </div>
            <h1
              class="text-2xl"
            >
               
              10kph
               
            </h1>
          </div>
          <div
            class="flex justify-start"
          >
            <div
              class="flex w-40 items-center gap-2 text-center text-3xl"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 2c-5.33 4.55-8 8.48-8 11.8 0 4.98 3.8 8.2 8 8.2s8-3.22 8-8.2c0-3.32-2.67-7.25-8-11.8zM7.83 14c.37 0 .67.26.74.62.41 2.22 2.28 2.98 3.64 2.87.43-.02.79.32.79.75 0 .4-.32.73-.72.75-2.13.13-4.62-1.09-5.19-4.12a.75.75 0 01.74-.87z"
                />
              </svg>
              <span
                class="text-2xl"
              >
                Humidity
              </span>
            </div>
            <h1
              class="text-2xl"
            >
               
              50%
               
            </h1>
          </div>
          <div
            class="flex justify-start"
          >
            <div
              class="flex w-40 items-center gap-2 text-center text-3xl"
            >
              <svg
                fill="currentColor"
                height="1em"
                id="Layer_1"
                stroke="currentColor"
                stroke-width="0"
                version="1.1"
                viewBox="0 0 30 30"
                width="1em"
                x="0px"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <path
                  d="M7.69,13.2c0-0.99,0.19-1.94,0.58-2.85c0.39-0.91,0.91-1.68,1.57-2.33s1.44-1.17,2.34-1.56c0.9-0.39,1.85-0.58,2.84-0.58
	c0.99,0,1.94,0.19,2.85,0.58c0.9,0.39,1.68,0.91,2.33,1.56c0.65,0.65,1.17,1.43,1.56,2.33s0.58,1.85,0.58,2.85
	c0,1.62-0.48,3.06-1.44,4.34c-0.96,1.27-2.2,2.14-3.71,2.61v3.29h-4.24v-3.25c-1.54-0.45-2.81-1.32-3.79-2.61S7.69,14.83,7.69,13.2z
	 M9.3,13.2c0,1.55,0.56,2.88,1.69,3.99c1.11,1.12,2.45,1.68,4.02,1.68c1.03,0,1.99-0.25,2.86-0.76c0.88-0.51,1.57-1.2,2.09-2.07
	c0.51-0.87,0.77-1.82,0.77-2.85c0-0.77-0.15-1.5-0.45-2.21s-0.71-1.31-1.22-1.82c-0.51-0.51-1.12-0.92-1.83-1.22
	c-0.71-0.3-1.44-0.45-2.21-0.45c-0.77,0-1.5,0.15-2.21,0.45s-1.31,0.71-1.82,1.22c-0.51,0.51-0.92,1.12-1.22,1.82
	C9.45,11.7,9.3,12.43,9.3,13.2z M9.88,13.56v-0.72h2.17v0.72H9.88z M10.97,10.02l0.52-0.52l1.52,1.52l-0.52,0.53L10.97,10.02z
	 M13.5,14.95c0-0.42,0.15-0.78,0.44-1.09c0.29-0.31,0.65-0.47,1.06-0.48l2.73-4.49l0.66,0.35l-2.02,4.83
	c0.18,0.25,0.26,0.54,0.26,0.88c0,0.44-0.15,0.81-0.46,1.11c-0.31,0.3-0.68,0.45-1.12,0.45c-0.43,0-0.8-0.15-1.1-0.45
	C13.65,15.76,13.5,15.39,13.5,14.95z M14.81,10.28V8.12h0.69v2.17H14.81z M17.75,13.55v-0.74h2.17v0.74H17.75z"
                />
              </svg>
              <span
                class="text-2xl"
              >
                Pressure
              </span>
            </div>
            <h1
              class="text-2xl"
            >
               
              1010hPa
               
            </h1>
          </div>
        </div>
      </section>
      <section
        class="flex flex-col gap-2"
      >
        <h1
          class="text-2xl font-semibold"
        >
          Next 
          0
           hours
        </h1>
        <div
          class="flex flex-wrap justify-center gap-5"
        />
      </section>
    </div>
  </div>
</div>
`;
